packages/leafygreen-provider/src/OverlayContext/OverlayContext.tsx

```diff
@@ -0,0 +1,79 @@
+import React, {
+  createContext,
+  ReactNode,
+  RefObject,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import PropTypes from 'prop-types';
+
+interface OverlayItem<E extends HTMLElement = HTMLElement> {
+  element: E;
+  ref: RefObject<E>;
+  id: string;
+}
+
+export interface OverlayContextProps {
+  getOverlays: () => Array<OverlayItem>;
+  registerOverlay: (overlay: OverlayItem) => void;
+  removeOverlay: (idToRemove: string) => void;
+  topMostOverlay: OverlayItem | undefined;
+}
+
+export const OverlayContext = createContext<OverlayContextProps | undefined>(
+  undefined,
+);
+
+export const useOverlayContext = (): OverlayContextProps => {
+  const context = useContext(OverlayContext);
+
+  if (!context) {
+    throw new Error('useOverlayContext must be used within an OverlayProvider');
+  }
+
+  return context;
+};
+
+/**
+ * Global provider tracking overlays that stack on z-axis
+ */
+const OverlayProvider = ({ children }: { children: ReactNode }) => {
+  const [overlays, setOverlays] = useState<Array<OverlayItem>>([]);
+
+  const topMostOverlay = useMemo(
+    () => overlays[overlays.length - 1],
+    [overlays],
+  );
+
+  const registerOverlay = useCallback((overlay: OverlayItem) => {
+    if (overlays.findIndex(item => item.id === overlay.id) > -1) return;
```

Comment:
Since all we're doing is checking if any overlay exists with this ID, it may be more readable to just use `Array.prototype.some()` here.
```suggestion
    if (overlays.some(item => item.id === overlay.id)) return;
```